os/kernel/log_dump/log_dump.c

```diff
@@ -0,0 +1,172 @@
+#endif
+
+#define LOG_DUMP_OK			 0
+#define LOG_DUMP_MEM_FAIL		-1
+#define LOG_DUMP_HEAP_FAIL		-2
+#define LOG_DUMP_SIZE_FAIL		-3
+
+/****************************************************************************
+ * Private Type Declarations
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Variables
+ ****************************************************************************/
+
+sq_queue_t log_dump_chunks; /* queue for holding the logs */
+int curbytes = 0; /* number of bytes in the current chunk that were filled */
+
+/****************************************************************************
+ * Private Variables
+ ****************************************************************************/
+
+static bool log_dump_chunks_initialized = false;
+static size_t log_net_size = 0;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void log_dump_init(void) {
+	sq_init(&log_dump_chunks);
+}
+
+int add_char_to_log_net(char ch)
+{
+	if (sq_empty(&log_dump_chunks)) {
+		/* no chunks are present so we allocate new chunk */
+		struct log_net_chunk_s *node = (struct log_net_chunk_s *)kmm_malloc(sizeof(struct log_net_chunk_s));
+
+		if (node == NULL) {
+			return LOG_DUMP_MEM_FAIL;
+		}
+
+		curbytes = 1;
+		node->arr[0] = ch;
+		sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+		/* set the size of chunks to 1 chunk */
+		log_net_size = sizeof(struct log_net_chunk_s);
+	} else {
+		/* need to check if the current chunks size is over max_log_size or greater than x% of free heap */
+		struct mm_heap_s *heap = mm_get_heap(sq_peek(&log_dump_chunks)); /* get heap details using the head of the chunks linked list */
+
+		if (heap == NULL) {
+			return LOG_DUMP_HEAP_FAIL;
+		}
+
+		size_t free_size = heap->mm_heapsize - heap->total_alloc_size; /* get the free size */
+		size_t percent_free_size = (free_size * CONFIG_LOG_DUMP_MAX_FREE_HEAP) / 100;
+		size_t max_size = MIN(percent_free_size, CONFIG_LOG_DUMP_MAX_SIZE);
+
+		while (!sq_empty(&log_dump_chunks) && log_net_size > max_size) {
+			struct log_net_chunk_s *node = (struct log_net_chunk_s *)sq_remfirst(&log_dump_chunks);
+			kmm_free(node);
+			log_net_size -= sizeof(struct log_net_chunk_s);
+		}
+
+		struct log_net_chunk_s *log_net_tail = (struct log_net_chunk_s *)sq_tail(&log_dump_chunks);
+
+		if (log_net_tail == NULL || curbytes == CONFIG_LOG_DUMP_CHUNK_SIZE) { /* all chunks were removed or last chunk is full */
+
+			/* there is memory, so fetch a new buffer */
+			struct log_net_chunk_s *node = (struct log_net_chunk_s *)kmm_malloc(sizeof(struct log_net_chunk_s));
```

Comment:
You use `sizeof(struct log...)` at many places. How about using definition?
```
#define LOG_CHUNK_SIZE (sizeof(struct ....))
```
Then, you can easily change the structure name later.